src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java

```diff
@@ -528,18 +608,51 @@ public void addExperience(final double xp)
         final double workBuildingLevel = getWorkBuilding() == null ? 0 : getWorkBuilding().getBuildingLevel();
         final double bonusXp = workBuildingLevel * (1 + citizenHutLevel) / Math.log(this.getExperienceLevel() + 2.0D);
         localXp = localXp * bonusXp;
+        final double saturation = citizenData.getSaturation();
+
+        if (saturation < AVERAGE_SATURATION)
+        {
+            if (saturation <= 0)
+            {
+                return;
+            }
+
+            if (saturation < LOW_SATURATION)
+            {
+                localXp -= localXp * BIG_SATURATION_FACTOR * saturation;
+            }
+            else
+            {
+                localXp -= localXp * LOW_SATURATION_FACTOR * saturation;
+            }
+        }
+        else if (saturation > AVERAGE_SATURATION)
+        {
+            if (saturation > HIGH_SATURATION)
+            {
+                localXp += localXp * BIG_SATURATION_FACTOR * saturation;
+            }
+            else
+            {
+                localXp += localXp * LOW_SATURATION_FACTOR * saturation;
+            }
+        }
+
         if (localXp > maxValue)
         {
             localXp = maxValue;
         }
-        citizenData.addExperience(localXp);
-
-        while (ExperienceUtils.getXPNeededForNextLevel(citizenData.getLevel()) < citizenData.getExperience())
+        if(citizenData != null)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Change this condition so that it does not always evaluate to "true" [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2583)